### PR TITLE
Added observers for all post ops

### PR DESCRIPTION
**Pull request by santanusinha**
* Moved most of the functionality out of Action
* QueryExecutor minor refactor
* Added fromCache to response
* Added tests for testing caching behaviour

### CODE REVIEW

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/SlowQueryReporter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ */
+@Slf4j
+public class SlowQueryReporter implements ActionExecutionObserver {
+    @Override
+    public void handleExecution(ActionEvaluationResponse response) {
+        if(null != response.getException() || response.isCached()) {
+            return;
+        }
+        if (response.getElapsedTime() > 1000) {
```
Comment (r0goyal) [inline foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/SlowQueryReporter.java]: Maybe this can be made configurable per action ?

Comment (santanusinha) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/SlowQueryReporter.java]: Not in this merge request

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Action.java [hunk 2]
```diff
@@ -96,23 +80,17 @@ private void preProcessRequest(String email) {
...
-        cacheManager.getCacheFor(this.cacheToken)
-                .put(cacheKey, execute(parameter));
-        return cacheKey;
-    }
-
     public ActionValidationResponse validate(String email) {
         try {
             preProcessRequest(email);
-        } catch (MalformedQueryException e) {
+        }
```
[the remark this answers is not in the document]
Comment (santanusinha) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/common/Action.java]: Not here

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/SlowQueryReporter.java (hunk 1 above)
Comment (santanusinha) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/handlers/SlowQueryReporter.java]: Done

File: foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/ActionExecutionObserver.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.flipkart.foxtrot.core.querystore;
+
+/**
+ *
+ */
+public interface ActionExecutionObserver {
+    void handleExecution(ActionEvaluationResponse response);
```
[the remark this answers is not in the document]
Comment (santanusinha) [reply foxtrot-core/src/main/java/com/flipkart/foxtrot/core/querystore/ActionExecutionObserver.java]: Done.

**Pull request merged**